File: profiler/native/perfd/network/network_cache.cc
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "network_cache.h"

using std::lock_guard;
using std::mutex;
using std::vector;

namespace profiler {

NetworkCache::NetworkCache() : connections_(1000) {}

ConnectionDetails* NetworkCache::AddConnection(int64_t conn_id, int32_t app_id,
                                               int64_t start_timestamp) {
  if (connections_.full()) {
    // An old connection is about to get overwritten, so remove it from our map
    const auto& conn = connections_.Get(0);
    conn_id_map_.erase(conn.id);
  }

  ConnectionDetails new_conn;
  new_conn.id = conn_id;
  new_conn.app_id = app_id;
  new_conn.start_timestamp = start_timestamp;

  lock_guard<mutex> lock(connections_mutex_);
  // |Add| copies new_conn; instead of tracking the (temporary) original, make
  // sure we use the address of the *copy* instead.
  ConnectionDetails* conn_ptr = connections_.Add(new_conn);
  conn_id_map_[conn_id] = conn_ptr;
  return conn_ptr;
}

ConnectionDetails* NetworkCache::GetDetails(int64_t conn_id) {
  return DoGetDetails(conn_id);
}

const ConnectionDetails* NetworkCache::GetDetails(int64_t conn_id) const {
  return DoGetDetails(conn_id);
}

vector<ConnectionDetails> NetworkCache::GetRange(int32_t app_id, int64_t start,
                                                 int64_t end) const {
  lock_guard<mutex> lock(connections_mutex_);
  vector<ConnectionDetails> data_range;
  for (size_t i = 0; i < connections_.size(); ++i) {
    const auto& conn = connections_.Get(i);
    if (conn.app_id != app_id) continue;

    // Given a range t0 and t1 and requests a-f...
    //
    //               t0              t1
    // a: [===========|===============|=========...
    // b: [=======]   |               |
    // c:         [===|===]           |
    // d:             |   [=======]   |
    // e:             |           [===|===]
    // f:             |               |   [=======]
    //
    // Keep a, c, d, and e; exclude b and f

    if (end < conn.start_timestamp) {
      break;  // Eliminate requests like f (and all requests after)
    }

    // At this point: conn.start_timestamp <= end
    if (conn.end_timestamp != 0 && conn.end_timestamp < start) {
      continue;  // Eliminate requests like b
    }

    data_range.push_back(conn);
  }

  return data_range;
}

ConnectionDetails* NetworkCache::DoGetDetails(int64_t conn_id) const {
  lock_guard<mutex> lock(connections_mutex_);
  auto it = conn_id_map_.find(conn_id);
  if (it != conn_id_map_.end()) {
    return it->second;
  } else {
    return nullptr;
  }
}

}  // namespace profiler

File: sdk-common/src/main/java/com/android/ide/common/repository/NetworkCache.kt
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.repository

import com.google.common.io.Files
import java.io.BufferedInputStream
import java.io.ByteArrayInputStream
import java.io.File
import java.io.FileInputStream
import java.io.InputStream
import java.util.concurrent.TimeUnit

/**
 * Provides a basic network cache with local disk fallback for data read
 * from a URL.
 */
abstract class NetworkCache constructor(
    private val baseUrl: String,

    /** Key used in cache directories to locate the network cache */
    val cacheKey: String,

    /** Location to search for cached repository content files */
    val cacheDir: File? = null,

    /**
     * Number of milliseconds to wait until timing out attempting to access the remote
     * repository
     */
    private val networkTimeoutMs: Int = 3000,

    /** Maximum allowed age of cached data; default is 7 days */
    private val cacheExpiryHours: Int = TimeUnit.DAYS.toHours(7).toInt(),

    /** If false, this repository won't make any network requests */
    private val networkEnabled: Boolean = true
) {

    /** Reads the given query URL in, with the given time out, and returns the bytes found */
    protected abstract fun readUrlData(url: String, timeout: Int): ByteArray?

    /** Provides the data from offline/local storage, if possible */
    protected abstract fun readDefaultData(relative: String): InputStream?

    /** Reports an error found during I/O */
    protected abstract fun error(throwable: Throwable, message: String?)

    /** Reads the given data relative to the base URL. */
    protected open fun findData(relative: String): InputStream? {
        if (cacheDir != null) {
            synchronized(cacheDir) {
                val file = File(cacheDir, if (relative.isNotEmpty()) relative else cacheKey)
                if (file.exists()) {
                    val lastModified = file.lastModified()
                    val now = System.currentTimeMillis()
                    val expiryMs = TimeUnit.HOURS.toMillis(cacheExpiryHours.toLong())

                    if (lastModified == 0L || now - lastModified <= expiryMs) {
                        // We found a cached file. Make sure it's actually newer than what the IDE
                        // ships with? Not really necessary since within the cache expiry interval
                        // it will be refreshed anyway
                        return BufferedInputStream(FileInputStream(file))
                    }
                }

                if (networkEnabled) {
                    try {
                        val data = readUrlData(
                          "$baseUrl$relative",
                          networkTimeoutMs
                        )
                        if (data != null) {
                            val parent = file.parentFile
                            parent?.mkdirs()
                            Files.write(data, file)
                            return ByteArrayInputStream(data)
                        }
                    }
                    catch (e: AssertionError) {
                        // Make sure that we propagate assertions
                        throw e
                    }
                    catch (e: Throwable) {
                        // timeouts etc: fall through to use built-in data
                    }
                }
            }
        }

        // Fallback: Builtin index, used for offline scenarios etc
        return readDefaultData(relative)
    }
}
